src/bounded_loop.rs

```diff
@@ -0,0 +1,156 @@
+        LoopId(id)
+    }
+
+    pub fn next(&self) -> Self {
+        LoopId(self.0 + 1)
+    }
+
+    pub fn id(&self) -> usize {
+        self.0
+    }
+}
+
+impl fmt::Display for LoopId {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+/// LoopStep is used to record or impose a bounded limit to a specific loop, which is identified by an id.
+pub trait LoopStep {
+    fn step(&mut self);
+}
+
+/// LoopCount tracks and report the maximum bound for this specific loop.
+pub struct LoopCount<'a> {
+    /// identifies a specific loop
+    id: LoopId,
+    /// count keeps the track of the current execution of this loop
+    count: usize,
+    /// reference to a static allocated AtomicUsize which keeps track of
+    atomic: &'a AtomicUsize,
+}
+
+impl<'a> LoopStep for LoopCount<'a> {
+    /// step is called on each iteration of a loop
+    fn step(&mut self) {
+        self.count = self.count.saturating_add(1);
+    }
+}
+
+impl<'a> Drop for LoopCount<'a> {
+    fn drop(&mut self) {
+        // TODO: llogiq suggested to use atomic max as the following code is not accurate due to TOCTOU
+        let current = self.atomic.load(Ordering::SeqCst);
+
+        if self.count > current {
+            self.atomic.store(self.count, Ordering::SeqCst);
+
+            OpenOptions::new()
+                .create(true)
+                .append(true)
+                .truncate(false)
+                .open("target/mutagen/loops.txt")
+                .and_then(|mut f| {
+                    f.write_all(format!("{},{}\n", self.id, self.count).as_str().as_ref())
+                });
+        }
+    }
+}
+
+impl<'a> LoopCount<'a> {
+    /// Creates the LoopCounter on recording mode and it won't impose a maximum bound
+    pub fn new(id: LoopId, atomic: &'a AtomicUsize) -> Self {
+        LoopCount {
+            id,
+            count: 0,
+            atomic,
+        }
+    }
+}
+
+/// LoopBound keeps track of the amount of iterations of a loop and it will stop when a limit is
+/// reached. This limit is calculated from previous executions of the tests with mutation count == 0.
+/// Then we add some base amount of iterations and we multiply by a factor to give some margin.
+/// Finally, if the limit is reached, it exits the process.
+pub struct LoopBound {
+    id: LoopId,
+    count: usize,
+    bound: usize,
+}
+
+impl LoopBound {
+    /// Creates a `LoopStep` with a maximum bound
+    pub fn new(id: LoopId) -> Self {
+        let count = MUTAGEN_LOOP_COUNT.get(&id).unwrap_or(&0);
+        let bound = BoundCalculator::calculate(*count);
+
+        LoopBound {
+            id,
+            count: 0,
+            bound,
+        }
+    }
+}
+
+impl LoopStep for LoopBound {
+    /// step is called on each iteration of a loop
+    fn step(&mut self) {
+        self.count = self.count.saturating_add(1);
+
+        if self.bound == self.count {
+            ::std::process::exit(-2);
```

Comment:
I did this planning to show an special status on the runner, but I'm not sure it's worth to do it.

Reply:
I like the idea of being able to identify loop overflows from the runner, this would enable us to do better reporting.